feat/re-allow multiple workers (#36134)
open-telemetry/opentelemetry-collector-contrib: exporter/prometheusremotewriteexporter/README.md

- `max_batch_size_bytes` (default = `3000000` -> `~2.861 mb`): Maximum size of a batch of
  samples to be sent to the remote write endpoint. If the batch size is larger
  than this value, it will be split into multiple batches.
- `max_batch_request_parallelism` (default = `5`): Maximum parallelism allowed for a single request bigger than `max_batch_size_bytes`.

>   This configuration is used only when feature gate `exporter.prometheusremotewritexporter.EnableMultipleWorkers` is enabled.

Reviewer: I think the implementation differs from this documentation.  It seems like this config option will be used if set regardless of the state of the feature gate.

Reviewer 2: @Aneurysm9 is correct.  Please remove this sentence

Author: Correct. I didnt review README after all the changes. Please take another look. If it LGTM do let me know to squash and rebase

### Feature gates

#### RetryOn429

This exporter has feature gate: `exporter.prometheusremotewritexporter.RetryOn429`.
When this feature gate is enable the prometheus remote write exporter will retry on 429 http status code with the provided retry configuration.
It currently doesn't support respecting the http header `Retry-After` if provided since the retry library used doesn't support this feature.

To enable it run collector with enabled feature gate `exporter.prometheusremotewritexporter.RetryOn429`. This can be done by executing it with one additional parameter - `--feature-gates=telemetry.useOtelForInternalMetrics`.

#### EnableMultipleWorkersFeatureGate

This exporter has feature gate: `+exporter.prometheusremotewritexporter.EnableMultipleWorkers`.

When this feature gate is enabled, `num_consumers` will be used as the worker counter for handling batches from the queue, and `max_batch_request_parallelism` will be used for parallelism on single batch bigger than `max_batch_size_bytes`.

> Enabling this feature gate, with `num_consumers` higher than 1 requires the target destination to supports ingestion of OutOfOrder samples. See [Multiple Consumers and OutOfOrder](#multiple-consumers-and-outoforder) for more info

Reviewer: I think this value needs to default to `1` if this feature gate is enabled.  When the feature gate is ultimately removed I don't think it appropriate to have a default value that is known to not work with any number of potential receivers and would not work with the default configuration of those receivers that are known to be able to support it in some configurations.

Author: I aggree that will help with transition, and that the exporter should have "sane" defaults at all times. 
done. please check

[beta]:https://github.com/open-telemetry/opentelemetry-collector#beta
[contrib]:https://github.com/open-telemetry/opentelemetry-collector-releases/tree/main/distributions/otelcol-contrib
[core]:https://github.com/open-telemetry/opentelemetry-collector-releases/tree/main/distributions/otelcol

## Multiple Consumers and OutOfOrder

> **DISCLAIMER**: This snippet applies only to Prometheus, other remote write destinations using Prometheus Protocol (ex: Thanos/Grafana Mimir/VictoriaMetrics) may have different settings.
> 
> By default, Prometheus tipically expects samples to be ingested sequentially, in temporal order.
> 
> When multiple consumers are enabled, the temporal ordering the samples written to the target destination is not deterministic and temporal ordering can no longer be garanteed. Example: one worker may be pushing a sample for `t+30s` and a second worker may push an additional sample but for `t+15s`.
> 
> Vanilla Prometheus configurations will reject these unordered samples. You will find similar error to this `Error on ingesting out-of-order samples`.
> 
> For enabling multiple consumers, it is necessary to enable out-of-order support in Prometheus.
> This can be done by using the `tsdb.out_of_order_time_window: 10m` settings. Please choose an appropriate time window to support pushing the worst case scenarios of a "queue" build up on the sender side.

Reviewer 3: ```suggestion
**DISCLAIMER**: This snippet applies only to Prometheus, other remote write destinations using Prometheus Protocol (ex: Thanos/Grafana Mimir/VictoriaMetrics) may have different settings.

By default, Prometheus expects samples to be ingested sequentially, in temporal order.

When multiple consumers are enabled, the temporal ordering of the samples written to the target destination is not deterministic, and temporal ordering can no longer be guaranteed. For example, one worker may push a sample for `t+30s`, and a second worker may push an additional sample but for `t+15s`.

Vanilla Prometheus configurations will reject these unordered samples and you'll receive "out of order" errors.

Out-of-order support in Prometheus must be enabled for multiple consumers.
This can be done by using the `tsdb.out_of_order_time_window: 10m` settings. Please choose an appropriate time window to support pushing the worst-case scenarios of a "queue" build-up on the sender side.
```

## Multiple Consumers and OutOfOrder

**DISCLAIMER**: This snippet applies only to Prometheus, other remote write destinations using Prometheus Protocol (ex: Thanos/Grafana Mimir/VictoriaMetrics) may have different settings.

By default, Prometheus expects samples to be ingested sequentially, in temporal order.

When multiple consumers are enabled, the temporal ordering of the samples written to the target destination is not deterministic, and temporal ordering can no longer be guaranteed. For example, one worker may push a sample for `t+30s`, and a second worker may push an additional sample but for `t+15s`.

Vanilla Prometheus configurations will reject these unordered samples and you'll receive "out of order" errors.

Out-of-order support in Prometheus must be enabled for multiple consumers.
This can be done by using the `tsdb.out_of_order_time_window: 10m` settings. Please choose an appropriate time window to support pushing the worst-case scenarios of a "queue" build-up on the sender side.

See for more info:
- https://prometheus.io/docs/prometheus/latest/configuration/configuration/#tsdb

> - https://prometheus.io/docs/prometheus/latest/feature_flags/#remote-write-receiver

Reviewer 3: Could we remove it?

> - https://prometheus.io/docs/prometheus/latest/feature_flags/#remote-write-receiver

Author: upstream doc may have hanged. will remove this entry.

> - https://prometheus.io/docs/prometheus/latest/feature_flags/#remote-write-receiver

Author: done
